### ISSUE TITLE
set_seed does not make training runs reproducible, and nothing records that

### ISSUE THREAD

**Issue by magilliam27** [magilliam27/MCI-GRU#132]
## Summary

`set_seed` seeds RNG streams but does not make a training run reproducible, and until #130 the module claimed otherwise. There is currently no way to reproduce a training run bitwise, and nothing in a run's artifacts records that.

Surfaced while correcting the `mci_gru/utils/seeding.py` docstring in #130. No code change was made beyond the docstring; this issue records the gap.

## What `set_seed` does and does not do

```python
def set_seed(seed: int) -> None:
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(seed)
```

That seeds four RNG streams. It does **not** call `torch.use_deterministic_algorithms(True)`, and does not set `torch.backends.cudnn.deterministic` or `torch.backends.cudnn.benchmark`. Several PyTorch operations select nondeterministic kernels by default, and cuDNN autotuning may pick different algorithms between runs on identical inputs.

So two runs with the same seed can differ, and the difference is invisible: nothing errors, nothing warns, and the run artifacts record the seed as though it pinned the result.

Until #130 the module docstring read *"Deterministic RNG seeding for experiments and ensemble members"*, which asserted the guarantee it does not provide. #130 retracts the claim. It does not close the gap.

## Why this matters here specifically

This repository holds an unusually strict evidence standard — trial ledgers, mutation-checked tests, golden fixtures, resolved-config provenance from #110. Seeds are treated as part of experiment identity, and seed sweeps are an established research pattern here.

If a seed does not pin a result, then a seed recorded in a trial ledger identifies an intent rather than an outcome, and two ledger rows with the same seed are not necessarily comparable. That weakens exactly the artifact chain the programme has been building.

Worth distinguishing from what is already covered: `test_backtest_engine_deterministic` asserts that running the backtest CLI twice yields identical output. That is **replay** stability over saved predictions in one environment, and it passes. It says nothing about whether *training* reproduces.

## Bitwise-policy decision

The maintainer accepted deferring optional strict-determinism support until repeatability
measurements show whether it is needed. Bitwise training equality is not a required guarantee
for this milestone. Revisit stricter execution if same-seed variation materially changes
rankings or conclusions; do not widen tolerances merely to excuse instability. See
[the linked #142 decision](https://github.com/magilliam27/MCI-GRU/issues/142#issuecomment-5563322123).

The artifact and documentation acceptance criteria below remain outstanding. The historical
trade-offs that informed this decision are:

- `torch.use_deterministic_algorithms(True)` raises on operations lacking a deterministic implementation, so enabling it may simply fail until every such op is replaced.
- Deterministic cuDNN algorithms and disabled autotuning are typically slower.
- Determinism still does not survive a change of device, PyTorch version, or CUDA version, so the guarantee is narrower than it sounds even when enabled.

This may be a reasonable trade to decline. What is not reasonable is declining it silently.

## Possible directions

Current direction: record provenance under #142's accepted environment-equivalent target;
defer option 2 pending repeatability measurements. Option 3 alone does not deliver the
accepted full-provenance target. The alternatives considered were:

1. **Record, do not promise.** Capture the reproducibility-relevant environment — torch version, CUDA version, device, cuDNN flags, deterministic-algorithms state — into the resolved-config artifact from #110. Cheap, additive, and makes a run's comparability auditable after the fact. Does not make runs reproducible.
2. **Opt-in determinism mode.** A config flag that sets `torch.use_deterministic_algorithms(True)` and the cuDNN flags, accepting the failure and performance cost, for runs where reproducibility matters more than throughput.
3. **Document and stop.** Treat the #130 docstring as sufficient, and state in the testing guide that same-seed runs are not expected to be bitwise identical.

Option 1 is compatible with all three and is the smallest step that improves the evidence chain.

## Out of scope

- Changing default training behaviour or performance.
- Auditing whether past seed-sweep results were affected. That is a separate question and would need its own evidence.

## Acceptance criteria

- [x] A decision recorded on which direction to take.
- [ ] If option 1: a run's artifacts let a reader determine whether two runs were executed under comparable conditions, without inspecting the machine.
- [ ] If option 2: a mutation-checked test proving the flag actually changes the configured state, not merely that it is accepted.
- [ ] `docs/TESTING_GUIDE.md` states the actual guarantee, whatever is decided.

**Comment by magilliam27** [comment]
Policy outcome from the linked #142 owner exchange: [durable decision](https://github.com/magilliam27/MCI-GRU/issues/142#issuecomment-5563322123). The maintainer accepted environment-equivalent reruns with full provenance and measured tolerances, and has now accepted deferring optional strict-determinism support until repeatability measurements show whether it is needed. Bitwise training equality is not a required guarantee for this milestone. Revisit stricter execution if same-seed variation materially changes rankings or conclusions; widening tolerances to excuse instability is not the remedy.

This records the explicit bitwise-policy answer owed here. The ticket stays open: comparable-condition artifacts and the documented guarantee in `docs/TESTING_GUIDE.md` have not been delivered by this decision exchange. No training flags or default behavior were changed. #144 remains the provenance implementation owner, currently blocked on completion of #142's gate.
